### PR TITLE
Don't store subclass impl/private data in an Option<T>

### PR DESCRIPTION
**Pull request by sdroege**
We don't know the exact memory layout of the Option<T> in relation to T
so can't easily go from a *const T to the surrounding *const Option<T>.

Since nightly from 2020-02-12 using both pointer types interchangeably
causes segmentation faults as their memory layouts are not actually
compatible and never were, but due to some compiler changes this
actually causes crashes at runtime now.

See https://github.com/rust-lang/rust/issues/69102 for details.

As an Option<_> was only used for some paranoid runtime checks that are
not really needed, simply store T directly as impl/private data of
subclasses.

This has the side-effect of having zero-sized impl/private data types to
actually occupy zero bytes of memory, and in some other cases to save a
few bytes of the allocation.

----

CC @EPashkin @GuillaumeGomez

### CODE REVIEW

**Comment by sdroege** [comment]
@GuillaumeGomez This wants another bugfix release btw, it's a quite bad bug and I don't fully understand why it didn't explode from the very beginning :)

**Comment by GuillaumeGomez** [comment]
Indeed, thanks for coming up so quickly with this fix!

👍

**Comment by sdroege** [comment]
I have a small additional fix tomorrow, so don't do a release just yet. Tomorrow :)
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by GuillaumeGomez** [comment]
It's a bit late for a release you know... We still have a few details to discuss about as well!

File: src/subclass/object.rs [hunk 1]
```diff
@@ -564,6 +564,17 @@ mod test {
         assert!(weak.upgrade().is_none());
     }
 
+    #[test]
+    fn test_create_child_object() {
+        let type_ = ChildObject::get_type();
+        let obj = Object::new(type_, &[]).unwrap();
```
Comment (GuillaumeGomez) [inline src/subclass/object.rs]: Oh also, please use expect, even in test!

Comment (sdroege) [reply src/subclass/object.rs]: The tests in the whole file and in various other places are like this. Let's handle this in a separate PR to make it consistent

Comment (sdroege) [reply src/subclass/object.rs]: Created https://github.com/gtk-rs/glib/issues/592 for that.

**Comment by EPashkin** [comment]
👍 Thanks.

**Comment by sdroege** [comment]
Thanks for reviewing :)

**Comment by sdroege** [comment]
Released as 0.9.3 on crates.io

**Pull request merged**